### PR TITLE
Analyser/Generator: cleaning up of code

### LINKED ISSUE #1079: Analyser/Generator: cleaning up of code
As mentioned in the title, the idea is to clean up the code a bit especially with regards to:
 - branch coverage; and
 - check calls to `std::weak_ptr<T>::lock()`.

The latter is more generally addressed in issue #1078, but we still want to do it here in preparation for issue #882.

### PR DESCRIPTION
**Pull request by agarny**
Fixes #1079.

### CODE REVIEW

File: tests/analyser/analyser.cpp [hunk 1]
```diff
@@ -974,4 +980,11 @@ TEST(Analyser, coverage)
     ast->setValue({});
     ast->setVariable(libcellml::Variable::create());
     ast->setParent(libcellml::AnalyserEquationAst::create());
+
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("analyser/coverage.cellml"));
+    printIssues(parser);
+
+    analyser->analyseModel(model);
+    printIssues(analyser);
```
Comment (hsorby) [inline tests/analyser/analyser.cpp]: Is this for debugging purposes?

File: src/api/libcellml/analyservariable.h [hunk 2]
```diff
@@ -35,22 +35,18 @@ class LIBCELLML_EXPORT AnalyserVariable
...
-        CONSTANT,
-        COMPUTED_CONSTANT,
-        ALGEBRAIC,
-        EXTERNAL
+        VARIABLE_OF_INTEGRATION, /**< The variable is the variable of integration. */
+        STATE, /**< The variable is a state. */
+        CONSTANT, /**< The variable is a constant (e.g. x = 3). */
+        COMPUTED_CONSTANT, /**< The variable is a computed constant (e.g. x = 3+5, x = 3+z, x = y+z where y and z are constants). */
+        ALGEBRAIC, /**< The variable is an algebraic variable. */
+        EXTERNAL /**< The variable is an external variable. i.e. not computed as part of the model. */
```
Comment (hsorby) [inline src/api/libcellml/analyservariable.h]: ... variable, i.e. not ...

File: src/generator.cpp [hunk 3]
```diff
@@ -198,6 +226,15 @@ bool Generator::GeneratorImpl::modifiedProfile() const
                sha1(profileContents) != PYTHON_GENERATOR_PROFILE_SHA1;
 }
 
+std::string Generator::GeneratorImpl::newLineIfNeeded()
```
Comment (hsorby) [inline src/generator.cpp:215]: `newLineIfNotEmpty`
```
return mCode.empty() ? {} : "\n";
```

File: src/generator.cpp [hunk 4]
```diff
@@ -869,15 +815,19 @@ std::string Generator::GeneratorImpl::generateOperatorCode(const std::string &op
...
     auto astRightChild = ast->rightChild();
     auto astLeftChildCode = generateCode(astLeftChild);
     auto astRightChildCode = generateCode(astRightChild);
 
     // Determine whether parentheses should be added around the left and/or
-    // right piece of code, and this based on the precedence of the operators
-    // used in CellML, which are listed below from higher to lower precedence:
+    // right piece of code, and this based on the precedence of the
+    // operators used in CellML, which are listed below from higher to lower
+    // precedence:
```
Comment (hsorby) [inline src/generator.cpp]: Not sure about the width being used here for comments.  This has changed solely due to  a width setting of the user.

File: src/generator.cpp [hunk 5]
```diff
@@ -1195,119 +1151,99 @@ std::string Generator::GeneratorImpl::generatePiecewiseElseCode(const std::strin
...
 {
     // Generate the code for the given AST.
+    // Note: we don't handle the case of AnalyserEquationAst::Type::BVAR since
+    //       we don't need to generate any code for it.
 
     std::string code;
 
-    switch (ast->type()) {
-    case AnalyserEquationAst::Type::ASSIGNMENT:
+    if (ast->type() == AnalyserEquationAst::Type::ASSIGNMENT) {
```
Comment (hsorby) [inline src/generator.cpp]: Use a variable instead of calling the `ast->type()` method many times unnecessarily.

File: tests/analyser/analyserexternalvariable.cpp [hunk 6]
```diff
@@ -229,3 +229,22 @@ TEST(AnalyserExternalVariable, dependencyByName)
...
+    auto model = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));
+    auto other_model = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));
+
+    externalVariable = libcellml::AnalyserExternalVariable::create(model->component("membrane")->variable("V"));
+
+    externalVariable->addDependency(model->component("membrane")->variable("Cm"));
+
+    EXPECT_EQ(false, externalVariable->removeDependency(nullptr, "membrane", "Cm"));
+    EXPECT_EQ(false, externalVariable->removeDependency(model, "not_membrane", "Cm"));
+}
```
Comment (hsorby) [inline tests/analyser/analyserexternalvariable.cpp]: If this is solely for coverage purposes and not testing an aspect of the API it should be placed in the coverage test file.

File: tests/generator/generator.cpp [hunk 7]
```diff
@@ -2356,4 +2356,93 @@ TEST(Generator, coverage)
...
+
+    profile->setAlgebraicVariableTypeString("xxx");
+    profile->setExternalVariableTypeString("");
+
+    generator->implementationCode();
+
+    profile->setHasXorOperator(false);
+    profile->setXorFunctionString("");
+
+    generator->implementationCode();
```
Comment (hsorby) [inline tests/generator/generator.cpp]: Dump coverage tests for edge cases should be placed in the coverage file.

**Review by hsorby — CHANGES_REQUESTED** [review]

File: src/api/libcellml/analyservariable.h (hunk 2 above)
Comment (agarny) [reply src/api/libcellml/analyservariable.h]: Oops! Done in 3863ba32.

File: src/generator.cpp (hunk 3 above)
Comment (agarny) [reply src/generator.cpp:215]: Agreed. Done in 6a68a501, although it reads `return mCode.empty() ? "" : "\n"; ` since `{}` cannot be used on the right hand side of `?`.

File: src/generator.cpp (hunk 4 above)
Comment (agarny) [reply src/generator.cpp]: That code was modified at some point and unmodified and I missed that bit. Done in 6a68a501.

File: src/generator.cpp (hunk 5 above)
Comment (agarny) [reply src/generator.cpp]: Well, `AnalyserEquationAst::type()` simply returns `mPimpl->mType`, so I would expect the compiler to optimise that and make it inline, so not sure about creating a temporary variable just for that.

File: tests/analyser/analyser.cpp (hunk 1 above)
Comment (agarny) [reply tests/analyser/analyser.cpp]: Yes, it was. Clearly forgot to remove them! Done in 5612990f.

File: tests/analyser/analyserexternalvariable.cpp (hunk 6 above)
Comment (agarny) [reply tests/analyser/analyserexternalvariable.cpp]: I have seen that, but I am not sure I like it. When I work on something (e.g., Analyser or Generator), I like the idea of having all the tests related to it in the same place rather than scattered around.

**Comment by agarny** [comment]
Thanks @hsorby for the review, but remember that this PR relies on PR #1077. So, #1077 should first be merged in before we can merge this one in.

File: src/README.rst [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+   #ifdef LLVM_COVERAGE
+   #    define VALID_SHARED_PTR(ptr) (true)
+   #else
+   #    define VALID_SHARED_PTR(ptr) (ptr != nullptr)
+   #endif
+
+It works fine, but... it's a C macro and we really don't want to go that route, not to mention that it relies on conditional compilation which is not great either.
+Yet, we want 100% branch coverage with ``llvm-cov``.
+So, instead, we have decided not to test for ``sharedPtr != nullptr``.
```
Comment (nickerso) [inline src/README.rst]: Not a show-stopped for this PR, but this seems like information that should be in the developer guides not the repository README...and could be much less verbose.

File: src/analyser.cpp [hunk 9]
```diff
@@ -1213,37 +1240,43 @@ void Analyser::AnalyserImpl::analyseEquationAst(const AnalyserEquationAstPtr &as
...
                     }
-
-                    break;
                 }
-            }
-        } else if (!mModel->areEquivalentVariables(variable, mModel->mPimpl->mVoi->variable())) {
-            auto issue = Issue::IssueImpl::create();
+            } while (voi == nullptr);
+        } else {
+            // Note: see the llvm-cov section in src/README.rst for voiVariable.
```
Comment (nickerso) [inline src/analyser.cpp]: this seems unnecessary.

Comment (nickerso) [reply src/analyser.cpp]: same for all the other added instances of the same comment...don't think its a good idea to introduce cross-references like this as comments unless it really adds value.

File: src/analyserequation.cpp [hunk 10]
```diff
@@ -16,7 +16,10 @@ limitations under the License.
 
 #include "libcellml/analyserequation.h"
 
+#include <algorithm>
+
 #include "analyserequation_p.h"
+#include "commonutils.h"
```
Comment (nickerso) [inline src/analyserequation.cpp]: is this needed?

File: src/api/libcellml/analyserequationast.h [hunk 11]
```diff
@@ -38,177 +38,104 @@ class LIBCELLML_EXPORT AnalyserEquationAst
...
-        NEQ,
-        LT,
-        LEQ,
-        GT,
-        GEQ,
-        AND,
-        OR,
-        XOR,
-        NOT,
+        EQ, /**< The equal to function. */
```
Comment (nickerso) [inline src/api/libcellml/analyserequationast.h]: these are operators, not functions...right? (but I can see this is just improving the structure of the existing docs, so its not a show stopper)

**Review by nickerso — COMMENTED** [review]
I think all the comments added ` // Note: see the llvm-cov section in src/README.rst.` are unnecessary and should be removed.

(Review in progress but not sure when I'll get back to it, so posting comments now to see what others think..)

**Comment by agarny** [comment]
> I think all the comments added ` // Note: see the llvm-cov section in src/README.rst.` are unnecessary and should be removed.
> 
> (Review in progress but not sure when I'll get back to it, so posting comments now to see what others think..)

I believe this is something that @hsorby requested in one of our face-to-face meetings, but I would have no issue with removing them.

File: src/README.rst (hunk 8 above)
Comment (agarny) [reply src/README.rst]: Developer guide... which most people don't read? I don't even know / remember where it is.

I agree that it's a bit verbose and that's the reason it's in a dedicated `README.rst` file where I thought I would put as much information as possible so that the next time we go through the code we don't end up scratching our heads about why we did things the way we did them.

File: src/analyserequation.cpp (hunk 10 above)
Comment (agarny) [reply src/analyserequation.cpp]: Oops, it must have been needed at some point during my work on this PR, but it's clearly not needed anymore. Fixed in 74e5c3c.

File: src/api/libcellml/analyserequationast.h (hunk 11 above)
Comment (agarny) [reply src/api/libcellml/analyserequationast.h]: Oops. Fixed in 103abc16.

File: tests/analyser/analyserexternalvariable.cpp (hunk 6 above)
Comment (nickerso) [reply tests/analyser/analyserexternalvariable.cpp]: Just to note that I prefer to keep these tests added purely for coverage to the coverage test file. Makes it nice and clear what they are and why they exist compared to the normal tests which are often the best documentation around for how to use libCellML.

Comment (agarny) [reply tests/analyser/analyserexternalvariable.cpp]: I really don't like it. When I work on, say, the analyser, I run the analyser tests which included the coverage tests for the analyser. Now, I will need to run both the analyser tests and the coverage tests, not to mention that the coverage tests will run other coverage-related tests that are not relevant to my work. So, yeah, it doesn't make sense to me to scatter things around like that.

But 2-1, so here goes: b1d72b8. 😕

Now, surely, we want to be consistent, so we will have to do the same for some annotator, importer, and printer tests.

File: tests/generator/generator.cpp (hunk 7 above)
Comment (agarny) [reply tests/generator/generator.cpp]: Done in b1d72b8.

**Comment by hsorby** [comment]
> > I think all the comments added ` // Note: see the llvm-cov section in src/README.rst.` are unnecessary and should be removed.
> > (Review in progress but not sure when I'll get back to it, so posting comments now to see what others think..)
> 
> I believe this is something that @hsorby requested in one of our face-to-face meetings, but I would have no issue with removing them.

We did talk about this in our meeting, I thought it would be a good idea for finding locations where we have removed branches because the current testing regime doesn't trigger them.  But, if it is 2-1 for the removal then they should go.

File: src/README.rst (hunk 8 above)
Comment (hsorby) [reply src/README.rst]: I'm with @nickerso  on this, this is not the place for this information.  The developer guides come from the `docs` directory, literally one directory away from where this is now.  I'm sure you can find it.

Comment (agarny) [reply src/README.rst]: My bad, I had clearly never realised that they were in the same repo. Addressed in 426cefe0.

File: src/analyser.cpp (hunk 9 above)
Comment (agarny) [reply src/analyser.cpp]: Addressed in 426cefe0.

File: src/generator.cpp (hunk 5 above)
Comment (hsorby) [reply src/generator.cpp]: I have had a look at the assembly produced on my mac, as far as I can tell (my assembly knowledge is not super-awesome) the function call is not getting inlined.  I can share the assembly I looked at to come to this conclusion, if you would like to see if you reach the same conclusion.

File: src/analyser.cpp [hunk 12]
```diff
@@ -170,7 +170,12 @@ struct AnalyserInternalEquation
     static bool isKnownOdeVariable(const AnalyserInternalVariablePtr &odeVariable);
 
     static bool hasKnownVariables(const std::vector<AnalyserInternalVariablePtr> &variables);
+    bool hasKnownVariables();
+
+    static bool isNonConstantVariable(const AnalyserInternalVariablePtr &variable);
+
     static bool hasNonConstantVariables(const std::vector<AnalyserInternalVariablePtr> &variables);
+    bool hasNonConstantVariables();
```
Comment (hsorby) [inline src/analyser.cpp:178]: Is there a method to the spacing on these declarations?

File: src/generator.cpp (hunk 5 above)
Comment (agarny) [reply src/generator.cpp]: My assembly knowledge is anything goes back to COR, so... it's anything but current. :) I am therefore happy to take your word for it. Kind of disappointing of a modern compiler not to optimise that kind of call. I remember reading, many years ago, that such functions (which are dead simple) would be inlined by the compielr. Anyway, done in 356e1707.

File: src/analyser.cpp (hunk 12 above)
Comment (agarny) [reply src/analyser.cpp:178]: The first one is on its own because it's regarding known variables while the others are about non-constant variables. As for the second one, it's separated from the last two because it is an `isXxxx()` method while the others are `hasXxx()` methods.

**Review by hsorby — APPROVED** [review]

**Review by nickerso — APPROVED** [review]
